File: android/app/src/main/java/fr/myfidpass/core/auth/JwtAccessExpiry.kt
package fr.myfidpass.core.auth

import org.json.JSONObject
import java.util.Base64

object JwtAccessExpiry {
    fun expirationEpochSeconds(jwt: String): Long? {
        val parts = jwt.split('.')
        if (parts.size < 2) return null
        return runCatching {
            var payload = parts[1]
            val rem = payload.length % 4
            if (rem > 0) payload += "=".repeat(4 - rem)
            val decoded = Base64.getUrlDecoder().decode(payload)
            val json = JSONObject(String(decoded, Charsets.UTF_8))
            json.optLong("exp", 0L).takeIf { it > 0L }
        }.getOrNull()
    }

    fun stillWithinValidityWindow(token: String?, marginSeconds: Long = 30): Boolean {
        val jwt = token?.trim().orEmpty()
        if (jwt.isEmpty()) return false
        val exp = expirationEpochSeconds(jwt) ?: return false
        val now = System.currentTimeMillis() / 1000
        return exp - now > marginSeconds
    }

    fun shouldProactivelyRefresh(token: String?, withinSeconds: Long = 120): Boolean {
        val jwt = token?.trim().orEmpty()
        if (jwt.isEmpty()) return false
        val exp = expirationEpochSeconds(jwt) ?: return false
        val now = System.currentTimeMillis() / 1000
        return exp - now <= withinSeconds
    }
}

File: myfidpass/Services/Auth/JwtAccessExpiry.swift
//
//  JwtAccessExpiry.swift
//  myfidpass — aligné Android JwtAccessExpiry.kt
//

import Foundation

enum JwtAccessExpiry {
    static func expirationEpochSeconds(_ jwt: String) -> TimeInterval? {
        let parts = jwt.split(separator: ".")
        guard parts.count >= 2 else { return nil }
        var payload = String(parts[1])
        let rem = payload.count % 4
        if rem > 0 { payload += String(repeating: "=", count: 4 - rem) }
        guard let data = Data(base64Encoded: payload, options: .ignoreUnknownCharacters),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let exp = json["exp"] as? TimeInterval, exp > 0 else {
            return nil
        }
        return exp
    }

    static func stillWithinValidityWindow(_ token: String?, marginSeconds: TimeInterval = 30) -> Bool {
        guard let jwt = token?.trimmingCharacters(in: .whitespacesAndNewlines), !jwt.isEmpty,
              let exp = expirationEpochSeconds(jwt) else { return false }
        let now = Date().timeIntervalSince1970
        return exp - now > marginSeconds
    }

    static func shouldProactivelyRefresh(_ token: String?, withinSeconds: TimeInterval = 120) -> Bool {
        guard let jwt = token?.trimmingCharacters(in: .whitespacesAndNewlines), !jwt.isEmpty,
              let exp = expirationEpochSeconds(jwt) else { return false }
        let now = Date().timeIntervalSince1970
        return exp - now <= withinSeconds
    }
}
